[kube-proxy] Move UDP conntrack operations together to pkg/proxy/util/conntrack.go

m1093782566 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Fix TODO in pkg/proxy/iptables.go, see

https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L1632

Move UDP conntrack operations together to from `pkg/proxy/iptables/proxier.go` to `pkg/proxy/util/conntrack.go` so that make them more consistent and add some UTs.

**Which issue this PR fixes**

Fixes #49477

**Special notes for your reviewer**:

```release-note
NONE
```

Commit:
move UDP conntrack operations together to pkg/proxy/util/conntrack.go

kevin-wangzefeng commented:
/ok-to-test

m1093782566 commented:
@kubernetes/sig-network-pr-reviews

m1093782566 commented:
/test pull-kubernetes-e2e-gce-etcd3

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,34 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+// The solution is clearing the conntrack. Known issues:
+// https://github.com/docker/docker/issues/8795
+// https://github.com/kubernetes/kubernetes/issues/31983
+func ClearUDPConntrackForPort(execer exec.Interface, port int) {
```
This should return the error, not log.

m1093782566 replied:
Yes, it should return the error.

BTW, I find the existing DeleteServiceConnections() doesn't return the error, see

https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/util/conntrack.go#L34-L45

Should I fix them together?

m1093782566 replied:
I fixed it as well.

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,34 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+
+// DeleteServiceConnection uses the conntrack tool to delete the conntrack entries
+// for the UDP connections specified by the {origin, dest} IP pair.
+func DeleteUDPConnections(execer exec.Interface, origin, dest string) {
```
This should return the error, not log.

thockin replied:
Should be named ClearUDPConntrackForPeers or something?

m1093782566 replied:
> This should return the error, not log.

Yes, fixed.

> Should be named ClearUDPConntrackForPeers or something?

Yes, I agree `ClearUDPConntrackForPeers` should be a more descriptive name. Thanks!

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,34 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+func ClearUDPConntrackForPort(execer exec.Interface, port int) {
+	glog.V(2).Infof("Deleting conntrack entries for udp connections")
+	if port > 0 {
+		err := ExecConntrackTool(execer, "-D", "-p", "udp", "--dport", strconv.Itoa(port))
```
Does this do ALL IPs?  That doesn't seem right...

m1093782566 replied:
Yes, it do ALL IPs.

I am not sure I fully understand what you expect. Currently I just move the existing code block from iptables/proxier.go to util/conntrack.go.

thockin replied:
I'm just observing that this will clear conntrack record it maybe should not.  We might have multiple services using the same port, and this will clear conntrack for all of them.

E.g. 
- client A -> service A port 12345
- client B -> service B port 12345
- this runs for 12345
- this clears both

It's not clear this is right.  What do you think?

m1093782566 replied:
Yes, when different services use the same port, it seems that this function will clear unexpected conntrack records.

This function will be called only for NodePort type service. So, if NodePort type service A and OTHER type service B use the same port, then error will occur.

thockin replied:
I'll file a bug for this.

m1093782566 replied:
I have seen the issue you created. I can help to with this.

Should I address this issue in this PR or should I send another PR?

m1093782566 commented:
/test pull-kubernetes-bazel

m1093782566 commented:
/test pull-kubernetes-e2e-kops-aws

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,36 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+// https://github.com/docker/docker/issues/8795
+// https://github.com/kubernetes/kubernetes/issues/31983
+func ClearUDPConntrackForPort(execer exec.Interface, port int) error {
+	glog.V(2).Infof("Deleting conntrack entries for udp connections")
```
don't glog in libraries, please

m1093782566 replied:
Okay, fixed. Thanks!

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,36 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+	if port > 0 {
+		err := ExecConntrackTool(execer, "-D", "-p", "udp", "--dport", strconv.Itoa(port))
+		if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
+			return fmt.Errorf("conntrack return with error: %v", err)
```
the error already says "conntrack command returned".  You can either add info ("can't clear conntrack for UDP %d: %v") or just return the error

m1093782566 replied:
Fixed.

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,36 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+// https://github.com/kubernetes/kubernetes/issues/31983
+func ClearUDPConntrackForPort(execer exec.Interface, port int) error {
+	glog.V(2).Infof("Deleting conntrack entries for udp connections")
+	if port > 0 {
```
flip this.  

```
if port < 0 {
  error
}
success path
```

m1093782566 replied:
Okay, fixed.

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +57,36 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+		// TODO: Better handling for deletion failure. When failure occur, stale udp connection may not get flushed.
+		// These stale udp connection will keep black hole traffic. Making this a best effort operation for now, since it
+		// is expensive to baby sit all udp connections to kubernetes services.
+		return fmt.Errorf("conntrack return with error: %v", err)
```
same as above - more useful or don't gift-wrap the error.

m1093782566 replied:
Fixed.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -959,22 +959,16 @@ func (esp *endpointServicePair) IPPart() string {
-				glog.Errorf("conntrack return with error: %v", err)
+			err := utilproxy.ClearUDPConntrackForPeers(proxier.exec, svcInfo.clusterIP.String(), endpointIP)
+			if err != nil {
+				glog.Errorf("%v", err)
```
sorry, it would be nice to log enough to know WHICH service

m1093782566 replied:
Fixed.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1373,7 +1367,10 @@ func (proxier *Proxier) syncProxyRules() {
-					proxier.clearUDPConntrackForPort(lp.port)
+					err := utilproxy.ClearUDPConntrackForPort(proxier.exec, lp.port)
+					if err != nil {
+						glog.Errorf("%v", err)
```
sorry, it would be nice to log enough to know WHICH port

m1093782566 replied:
Fixed.

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -31,17 +31,21 @@ const noConnectionToDelete = "0 flow entries have been deleted"
 		}
 	}
+	if len(errs) != 0 {
+		return utilerrors.NewAggregate(errs)
```
under what conditions would we have more than one error from this function?

m1093782566 replied:
```
func DeleteServiceConnections(execer exec.Interface, svcIPs []string) error
```

The input of this function is IP slice, so this function will delete multiple connection conntracks. Current implementaion doesn't return immediately when an error occur. So, we probably will have more than on errors from this function.

Is there any problem?

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -56,3 +60,33 @@ func ExecConntrackTool(execer exec.Interface, parameters ...string) error {
+	}
+	err := ExecConntrackTool(execer, "-D", "-p", "udp", "--dport", strconv.Itoa(port))
+	if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
+		return fmt.Errorf("Failed to delete conntrack entries for UDP port: %d, error: %v", port, err)
```
Error strings should not start with capital letters - same everywhere (unfortunately, logs and errors are different on this :(

m1093782566 replied:
Fixed.

m1093782566 commented:
@thockin 

Thanks for your feedback. I fixed all comments except returning multiple error fun func `DeleteServiceConnections(execer exec.Interface, svcIPs []string) error`. I explained it inline. PTAL.

BTW, I added a TODO for issue: #49881

m1093782566 commented:
/test pull-kubernetes-e2e-gce-etcd3

thockin commented on pkg/proxy/util/conntrack.go:
```diff
@@ -31,17 +31,21 @@ const noConnectionToDelete = "0 flow entries have been deleted"
-func DeleteServiceConnections(execer exec.Interface, svcIPs []string) {
+func DeleteServiceConnections(execer exec.Interface, svcIPs []string) error {
+	var errs []error
 	for _, ip := range svcIPs {
```
Truthfully, this should probably be looping in the caller and this function should just run a single conntrack, so the caller can handle multiple errors cleanly.

m1093782566 replied:
I Fixed it by refactoring `DeleteServiceConnections()` only handle one service IP at a time. PTAL.

thockin reviewed (commented):
one point, then LGTM

m1093782566 commented:
/test pull-kubernetes-unit

m1093782566 commented:
/test pull-kubernetes-verify

m1093782566 commented:
/test pull-kubernetes-e2e-kops-aws

m1093782566 commented:
/test pull-kubernetes-unit

m1093782566 commented:
/test pull-kubernetes-unit

m1093782566 commented:
@thockin 

All CI checks are green now, would you please take a look again?

m1093782566 commented:
ping @freehan 

Do you have bandwidth to review this PR, thanks!

freehan commented on pkg/proxy/util/conntrack.go:
```diff
@@ -18,30 +18,27 @@ package util
-		}
+// DeleteServiceConnections uses the conntrack tool to delete the conntrack entries
+// for the UDP connections specified by the given service IP
+func DeleteServiceConnections(execer exec.Interface, ip string) error {
```
nit: maybe rename to ClearUDPConntrackForClusterIP?

m1093782566 replied:
Glad to receive your feedback. @freehan 

 I agree this function needs a better name. Since the parameter of this function is not limited to Cluster IP, so can we rename it to `ClearUDPConntrackForIP`? 

Please let me know if you have better ideas :) Thanks!

m1093782566 replied:
Fixed now. DeleteServiceConnections -> ClearUDPConntrackForIP

freehan reviewed (commented):
LGTM overall. Just one nit

m1093782566 commented:
@freehan 

Comments are fixed now. Please take a look when you have a chance. Thanks!

m1093782566 commented:
/test pull-kubernetes-unit

m1093782566 commented:
ping @thockin @freehan

freehan commented:
/lgtm
/approve

m1093782566 commented:
/test pull-kubernetes-federation-e2e-gce

m1093782566 commented:
ping @thockin 

This PR needs your approve. PTAL. Much appreciate.

thockin commented:
/approve

m1093782566 commented:
@thockin @freehan 

This PR is rebased due to conflicts, would you please take a look again and approve it? Much appreciate :)

freehan commented:
/lgtm
